### PR TITLE
feat: set kernel of support vector machine

### LINKED ISSUE #172: Set kernel of `SupportVectorMachine`
### Is your feature request related to a problem?

It's not possible to set the kernel of a `SupportVectorMachine`.

### Desired solution

* Add a new parameter `kernel: SupportVectorMachineKernel` to the initializer of `safeds.ml.classification.SupportVectorMachine` and `safeds.ml.regression.SupportVectorMachine`
* Pass it along as the `kernel` of the wrapped `scikit-learn` model in the `fit` method

The implementation of `SupportVectorMachineKernel` should be similar to the `ImputerStrategy`:

* `SupportVectorMachineKernel` should be an abstract base class
* `Kernel` should be nested into `SupportVectorMachine`
* Subclasses should be nested into `Kernel`
* Subclasses should be:
  * `Linear`
  * `Polynomial` (with a `degree: int` parameter)
    * Raise if `degree` < 1 
  * `Sigmoid`
  * `RadialBasisFunction`

Also add a getter as described in #260.

### Possible alternatives (optional)

_No response_

### Screenshots (optional)

_No response_

### Additional Context (optional)

We need to add the support vector machine first (#154).

### PR DESCRIPTION
**Pull request by jxnior01**
Closes #172.

A new parameter kernel: SupportVectorMachineKernel was added to the initializer of SupportVectorMachine classes of both the classifier and the regressor alongside getters for the parameters. It was later passed as the kernel of the wrapped scikit-learn model in the fit method. 
The SupportVectorMachineKernel was created as an abstract base class from which the signatures of the methods in the nested subclasses of Kernel were the same.
Tests were adjusted to test the functionality of the kernel parameter.

### CODE REVIEW

File: src/safeds/ml/classical/classification/_support_vector_machine.py [hunk 1]
```diff
@@ -1,19 +1,38 @@
...
     from safeds.data.tabular.containers import Table, TaggedTable
 
 
+class SupportVectorMachineKernel(ABC):
+    """The abstract base class of the different subclasses supported by the `Kernel`."""
+
+    @abstractmethod
+    def get_sklearn_kernel(self, svm: SupportVectorMachine) -> object:
+        """
+        Get the kernel of the given SupportVectorMachine.git.
```
[the remark this answers is not in the document]
Comment (jxnior01) [reply src/safeds/ml/classical/classification/_support_vector_machine.py:23]: Yes it doesn't. It was a typo.

File: src/safeds/ml/classical/classification/_support_vector_machine.py [hunk 2]
```diff
@@ -22,14 +37,15 @@ class SupportVectorMachine(Classifier):
     ----------
     c: float
         The strength of regularization. Must be strictly positive.
+        kernel: The type of kernel to be used. Defaults to None.
```
Comment (alex-senger) [inline src/safeds/ml/classical/classification/_support_vector_machine.py:40]:
```suggestion
    kernel: The type of kernel to be used. Defaults to None.
```

**Review by alex-senger — CHANGES_REQUESTED** [review]

File: src/safeds/ml/classical/regression/_support_vector_machine.py [hunk 3]
```diff
@@ -22,14 +37,15 @@ class SupportVectorMachine(Regressor):
     ----------
     c: float
         The strength of regularization. Must be strictly positive.
+        kernel: The type of kernel to be used. Defaults to None.
```
Comment (alex-senger) [inline src/safeds/ml/classical/regression/_support_vector_machine.py:40]:
```suggestion
    kernel: The type of kernel to be used. Defaults to None.
```

File: tests/safeds/ml/classical/classification/test_support_vector_machine.py [hunk 4]
```diff
@@ -11,13 +11,64 @@ def training_set() -> TaggedTable:
...
+        assert svm._get_kernel_name() == "sigmoid"
+
+        svm = SupportVectorMachine(c=2, kernel=SupportVectorMachine.Kernel.RadialBasisFunction())
+        assert svm._get_kernel_name() == "rbf"
+
+    def test_should_get_kernel_name_invalid_kernel_type(self) -> None:
+        svm = SupportVectorMachine(c=2)
+        svm._kernel = None
+        with pytest.raises(TypeError, match="Invalid kernel type."):
+            svm._get_kernel_name()
```
Comment (alex-senger) [inline tests/safeds/ml/classical/classification/test_support_vector_machine.py:71]: The Tests for the Kernel shouldn't be inside the TestC Class. Maybe rather add another Class named TestKernel.

File: tests/safeds/ml/classical/classification/test_support_vector_machine.py [hunk 5]
```diff
@@ -11,13 +11,64 @@ def training_set() -> TaggedTable:
...
+
+        svm = SupportVectorMachine(c=2, kernel=SupportVectorMachine.Kernel.Sigmoid())
+        assert svm._get_kernel_name() == "sigmoid"
+
+        svm = SupportVectorMachine(c=2, kernel=SupportVectorMachine.Kernel.RadialBasisFunction())
+        assert svm._get_kernel_name() == "rbf"
+
+    def test_should_get_kernel_name_invalid_kernel_type(self) -> None:
+        svm = SupportVectorMachine(c=2)
+        svm._kernel = None
```
Comment (alex-senger) [inline tests/safeds/ml/classical/classification/test_support_vector_machine.py:69]: Does the Test still work without this line?

File: tests/safeds/ml/classical/regression/test_support_vector_machine.py [hunk 6]
```diff
@@ -11,13 +11,64 @@ def training_set() -> TaggedTable:
...
+        assert svm._get_kernel_name() == "sigmoid"
+
+        svm = SupportVectorMachine(c=2, kernel=SupportVectorMachine.Kernel.RadialBasisFunction())
+        assert svm._get_kernel_name() == "rbf"
+
+    def test_should_get_kernel_name_invalid_kernel_type(self) -> None:
+        svm = SupportVectorMachine(c=2)
+        svm._kernel = None
+        with pytest.raises(TypeError, match="Invalid kernel type."):
+            svm._get_kernel_name()
```
Comment (alex-senger) [inline tests/safeds/ml/classical/regression/test_support_vector_machine.py:71]: Same here. Put these into a separate Class.

**Review by alex-senger — CHANGES_REQUESTED** [review]

File: tests/safeds/ml/classical/classification/test_support_vector_machine.py (hunk 5 above)
Comment (jxnior01) [reply tests/safeds/ml/classical/classification/test_support_vector_machine.py]: Yes it would. Line was written just to make clear, what was going on there.

File: tests/safeds/ml/classical/classification/test_support_vector_machine.py [hunk 7]
```diff
@@ -11,17 +11,69 @@ def training_set() -> TaggedTable:
 
 class TestC:
     def test_should_be_passed_to_fitted_model(self, training_set: TaggedTable) -> None:
-        fitted_model = SupportVectorMachine(c=2).fit(training_set=training_set)
+        kernel = SupportVectorMachine.Kernel.Linear()
+        fitted_model = SupportVectorMachine(c=2, kernel=kernel).fit(training_set=training_set)
```
Comment (alex-senger) [inline tests/safeds/ml/classical/classification/test_support_vector_machine.py:15]: Keep the Kernel as None. This method should only test the C Parameter and it would fail if something would be wrong with the Kernel implementation.

```suggestion
        fitted_model = SupportVectorMachine(c=2).fit(training_set=training_set)
```

File: tests/safeds/ml/classical/classification/test_support_vector_machine.py [hunk 8]
```diff
@@ -11,17 +11,69 @@ def training_set() -> TaggedTable:
...
         assert fitted_model.c == 2
+        assert isinstance(fitted_model.kernel, SupportVectorMachine.Kernel.Linear)
```
Comment (alex-senger) [inline tests/safeds/ml/classical/classification/test_support_vector_machine.py:17]: remove this line (see reason above)

File: tests/safeds/ml/classical/classification/test_support_vector_machine.py [hunk 9]
```diff
@@ -11,17 +11,69 @@ def training_set() -> TaggedTable:
...
 
     def test_should_be_passed_to_sklearn(self, training_set: TaggedTable) -> None:
-        fitted_model = SupportVectorMachine(c=2).fit(training_set)
+        kernel = SupportVectorMachine.Kernel.Linear()
+        fitted_model = SupportVectorMachine(c=2, kernel=kernel).fit(training_set)
         assert fitted_model._wrapped_classifier is not None
         assert fitted_model._wrapped_classifier.C == 2
+        assert isinstance(fitted_model.kernel, SupportVectorMachine.Kernel.Linear)
```
Comment (alex-senger) [inline tests/safeds/ml/classical/classification/test_support_vector_machine.py:24]: same here

File: tests/safeds/ml/classical/regression/test_support_vector_machine.py [hunk 10]
```diff
@@ -11,17 +11,69 @@ def training_set() -> TaggedTable:
...
         assert fitted_model.c == 2
+        assert isinstance(fitted_model.kernel, SupportVectorMachine.Kernel.Linear)
 
     def test_should_be_passed_to_sklearn(self, training_set: TaggedTable) -> None:
-        fitted_model = SupportVectorMachine(c=2).fit(training_set)
+        kernel = SupportVectorMachine.Kernel.Linear()
+        fitted_model = SupportVectorMachine(c=2, kernel=kernel).fit(training_set)
         assert fitted_model._wrapped_regressor is not None
         assert fitted_model._wrapped_regressor.C == 2
+        assert isinstance(fitted_model.kernel, SupportVectorMachine.Kernel.Linear)
```
Comment (alex-senger) [inline tests/safeds/ml/classical/regression/test_support_vector_machine.py:24]: same as above

**Review by alex-senger — CHANGES_REQUESTED** [review]

File: src/safeds/ml/classical/classification/_support_vector_machine.py [hunk 11]
```diff
@@ -1,19 +1,34 @@
...
 
+from abc import ABC, abstractmethod
 from typing import TYPE_CHECKING
 
 from sklearn.svm import SVC as sk_SVC  # noqa: N811
 
 from safeds.ml.classical._util_sklearn import fit, predict
-
-from ._classifier import Classifier
+from safeds.ml.classical.classification._classifier import Classifier
```
Comment (Marsmaennchen221) [inline src/safeds/ml/classical/classification/_support_vector_machine.py:9]:
```suggestion
from safeds.ml.classical.classification import Classifier
```

File: src/safeds/ml/classical/regression/_support_vector_machine.py [hunk 12]
```diff
@@ -1,19 +1,34 @@
...
 
+from abc import ABC, abstractmethod
 from typing import TYPE_CHECKING
 
 from sklearn.svm import SVR as sk_SVR  # noqa: N811
 
 from safeds.ml.classical._util_sklearn import fit, predict
-
-from ._regressor import Regressor
+from safeds.ml.classical.regression._regressor import Regressor
```
Comment (Marsmaennchen221) [inline src/safeds/ml/classical/regression/_support_vector_machine.py:9]:
```suggestion
from safeds.ml.classical.regression import Regressor
```

File: tests/safeds/ml/classical/classification/test_support_vector_machine.py (hunk 7 above)
Comment (jxnior01) [reply tests/safeds/ml/classical/classification/test_support_vector_machine.py:15]: As discussed, it will be better to move the test methods to the TestKernel class and perform only the kernel checks. If not code-coverage won't reach a 100%.

**Review by Marsmaennchen221 — CHANGES_REQUESTED** [review]

File: tests/safeds/ml/classical/classification/test_support_vector_machine.py (hunk 8 above)
Comment (jxnior01) [reply tests/safeds/ml/classical/classification/test_support_vector_machine.py:17]: Better to create a duplicate method for TestKernel that tests only the kernel.

File: src/safeds/ml/classical/regression/_support_vector_machine.py [hunk 13]
```diff
@@ -1,19 +1,34 @@
...
+
+    @abstractmethod
+    def get_sklearn_kernel(self) -> object:
+        """
+        Get the kernel of the given SupportVectorMachine.
+
+        Returns
+        -------
+        object
+              The kernel of the SupportVectorMachine.
```
Comment (alex-senger) [inline src/safeds/ml/classical/regression/_support_vector_machine.py:28]:
```suggestion
          The kernel of the SupportVectorMachine.
```

File: src/safeds/ml/classical/classification/_support_vector_machine.py [hunk 14]
```diff
@@ -1,19 +1,34 @@
...
+
+    @abstractmethod
+    def get_sklearn_kernel(self) -> object:
+        """
+        Get the kernel of the given SupportVectorMachine.
+
+        Returns
+        -------
+        object
+              The kernel of the SupportVectorMachine.
```
Comment (alex-senger) [inline src/safeds/ml/classical/classification/_support_vector_machine.py:28]:
```suggestion
          The kernel of the SupportVectorMachine.
```

**Review by alex-senger — APPROVED** [review]
[LGTM](https://lgtm.party/p/104743R/)
